gunicorn/reloader.py

```diff
@@ -68,48 +70,46 @@ def run(self):
 
 if has_inotify:
 
-    class InotifyReloader(threading.Thread):
+    class InotifyReloader(ReloaderBase):
         event_mask = (inotify.constants.IN_CREATE | inotify.constants.IN_DELETE
                       | inotify.constants.IN_DELETE_SELF | inotify.constants.IN_MODIFY
                       | inotify.constants.IN_MOVE_SELF | inotify.constants.IN_MOVED_FROM
                       | inotify.constants.IN_MOVED_TO)
 
         def __init__(self, extra_files=None, callback=None):
-            super().__init__()
-            self.setDaemon(True)
-            self._callback = callback
+            super().__init__(extra_files=extra_files, callback=callback)
             self._dirs = set()
             self._watcher = Inotify()
 
-            for extra_file in extra_files:
-                self.add_extra_file(extra_file)
-
         def add_extra_file(self, filename):
             dirname = os.path.dirname(filename)
 
-            if dirname in self._dirs:
-                return
+            with self._extra_files_lock:
+                super(InotifyReloader, self).add_extra_file(filename)
 
-            self._watcher.add_watch(dirname, mask=self.event_mask)
-            self._dirs.add(dirname)
+                if dirname in self._dirs:
+                    return
+
+                self._watcher.add_watch(dirname, mask=self.event_mask)
+                self._dirs.add(dirname)
 
         def get_dirs(self):
-            fnames = [
-                os.path.dirname(COMPILED_EXT_RE.sub('py', module.__file__))
-                for module in tuple(sys.modules.values())
-                if getattr(module, '__file__', None)
-            ]
+            dirnames = [os.path.dirname(fname) for fname in self.get_files()]
+            return set(dirnames)
 
-            return set(fnames)
+        def refresh_dirs(self):
+            with self._extra_files_lock:
```

Comment:
Here also we do not need the log anymore. Gunicorn will call `add_extra_file` during initialization and then start the reloader. Once starter, only the reloader thread is expected to add more files / directories, so there is no lock.